File: solutions/scala/127-meeting-rooms-ii.scala
/**
 * Problem 127: Meeting Rooms II (LeetCode 253)
 * Difficulty: Med
 * Language: Scala
 */
def minMeetingRooms(intervals: List[(Int, Int)]): Int = {
  val starts = intervals.map(_._1).sorted
  val ends = intervals.map(_._2).sorted
  def go(ss: List[Int], es: List[Int], rooms: Int, maxR: Int): Int =
    ss match {
      case Nil => maxR
      case s :: srest => es match {
        case e :: erest if s < e => go(srest, es, rooms + 1, maxR max (rooms + 1))
        case _ :: erest => go(srest, erest, rooms, maxR)
      }
    }
  go(starts, ends, 0, 0)
}

File: solutions/ocaml/127-meeting-rooms-ii.ml
(*
  Problem 127: Meeting Rooms II (LeetCode 253)
  Difficulty: Med
  Language: OCaml
*)
let min_meeting_rooms intervals =
  let starts = List.map fst intervals |> List.sort compare in
  let ends = List.map snd intervals |> List.sort compare in
  let rec go ss es rooms max_r = match ss with
    | [] -> max_r
    | s :: srest -> match es with
      | e :: _ when s < e -> go srest es (rooms+1) (max max_r (rooms+1))
      | _ :: erest -> go srest erest rooms max_r
      | [] -> max_r
  in go starts ends 0 0
